File: example/Exam04BindService/app/src/main/java/com/gbox3d/exam04bindservice/MainActivity.java
package com.gbox3d.exam04bindservice;

import androidx.appcompat.app.AppCompatActivity;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    MyService myService;
    boolean isService = false; // 서비스 중인 확인용


    public final static String TAG = "exam04";

    ServiceConnection conn = new ServiceConnection() {
        public void onServiceConnected(ComponentName name,
                                       IBinder service) {
            // 서비스와 연결되었을 때 호출되는 메서드
            // 서비스 객체를 전역변수로 저장
            MyService.LocalBinder mb = (MyService.LocalBinder) service;
            myService = mb.getService(); // 서비스가 제공하는 메소드 호출하여
            // 서비스쪽 객체를 전달받을수 있슴
            isService = true;
        }

        public void onServiceDisconnected(ComponentName name) {
            // 서비스와 연결이 끊겼을 때 호출되는 메서드
            isService = false;
            Toast.makeText(getApplicationContext(),
                    "서비스 연결 해제",
                    Toast.LENGTH_LONG).show();
        }
    };


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ((Button)findViewById(R.id.btn_start)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(
                        MainActivity.this, // 현재 화면
                        MyService.class); // 다음넘어갈 컴퍼넌트

                bindService(intent, // intent 객체
                        conn, // 서비스와 연결에 대한 정의
                        Context.BIND_AUTO_CREATE);

            }
        });

        ((Button)findViewById(R.id.btn_stop)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(isService) {
                    unbindService(conn);
                    isService=false;
                }

            }
        });

        ((Button)findViewById(R.id.btn_test)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.d(TAG,"call bind test : " + myService.getNumber());

            }
        });

    }
}

File: ktPractice06_coroutin/app/src/main/java/com/example/ktpractice06_coroutin/MainActivity.kt
package com.example.ktpractice06_coroutin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        button_test_1.setOnClickListener() {

            GlobalScope.launch(Dispatchers.IO) {

                var _count = 0
                while (_count < 10) {
                    delay(500)
                    println("io done")
                    _count++
                    tv_test_1.text = "$_count"
                }

            }

        }

        button_test2.setOnClickListener() {
            GlobalScope.launch(Dispatchers.Main) {
                var _count = 0

                while (_count < 10) {
                    val _async = async(Dispatchers.IO) {
                        delay(500)
                        println("done io")
                        _count++
                    }

                    _async.await()
                    tv_test_2.text = "$_count"

                    println("done ui $_count")


                }

            }
        }
    }
}